src/components/structures/UserSettings.js

```diff
@@ -32,131 +40,227 @@ module.exports = React.createClass({
             avatarUrl: null,
             threePids: [],
             clientVersion: version,
-            phase: this.Phases.Loading,
+            phase: "UserSettings.LOADING", // LOADING, DISPLAY
         };
     },
 
     componentWillMount: function() {
         var self = this;
-        var cli = MatrixClientPeg.get();
-
-        var profile_d = cli.getProfileInfo(cli.credentials.userId);
-        var threepid_d = cli.getThreePids();
-
-        q.all([profile_d, threepid_d]).then(
-            function(resps) {
-                self.setState({
-                    avatarUrl: resps[0].avatar_url,
-                    threepids: resps[1].threepids,
-                    phase: self.Phases.Display,
-                });
-            },
-            function(err) { console.err(err); }
-        );
+        this._refreshFromServer();
     },
 
-    editAvatar: function() {
-        var url = MatrixClientPeg.get().mxcUrlToHttp(this.state.avatarUrl);
-        var ChangeAvatar = sdk.getComponent('settings.ChangeAvatar');
-        var avatarDialog = (
-            <div>
-                <ChangeAvatar initialAvatarUrl={url} />
-                <div className="mx_Dialog_buttons">
-                    <button onClick={this.onAvatarDialogCancel}>Cancel</button>
-                </div>
-            </div>
-        );
-        this.avatarDialog = Modal.createDialogWithElement(avatarDialog);
+    componentDidMount: function() {
+        this.dispatcherRef = dis.register(this.onAction);
+        this._me = MatrixClientPeg.get().credentials.userId;
     },
 
-    addEmail: function() {
+    componentWillUnmount: function() {
+        dis.unregister(this.dispatcherRef);
+    },
 
+    _refreshFromServer: function() {
+        var self = this;
+        q.all([
+            UserSettingsStore.loadProfileInfo(), UserSettingsStore.loadThreePids()
+        ]).done(function(resps) {
+            self.setState({
+                avatarUrl: resps[0].avatar_url,
+                threepids: resps[1].threepids,
+                phase: "UserSettings.DISPLAY",
+            });
+        }, function(error) {
+            var ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
```

Comment:
On 24/12/15 10:49, Kegsay wrote:
> In src/components/structures/UserSettings.js
> <https://github.com/matrix-org/matrix-react-sdk/pull/33#discussion_r48409055>:
>
>>
>> +    _refreshFromServer: function() {
>> +        var self = this;
>> +        q.all([
>> +            UserSettingsStore.loadProfileInfo(), UserSettingsStore.loadThreePids()
>> +        ]).done(function(resps) {
>> +            self.setState({
>> +                avatarUrl: resps[0].avatar_url,
>> +                threepids: resps[1].threepids,
>> +                phase: "UserSettings.DISPLAY",
>> +            });
>> +        }, function(error) {
>> +            var ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
>
> Hmm, maybe. We don't want to show the settings page. We probably don't
> want to show the spinner. ATM this will display the dialog and then
> continue spinning (clicking away and back to it with connection will
> refresh it, which is probably good enough for now given we don't have
> any assets for displaying global errors on the user settings page).

Just thinking that showing a spinner when nothing is happening is pretty 
horrid - and adding another state which doesn't show the settings or the 
spinner would be pretty trivial.